File: LeetCode/Apple/TopQuestions/Sorting_Searching/intersect.py
#Given 2 arrays, write a function to compute
#their intersection.


#Example 1:
#Input: nus1=[1,2,2,1]
#nums2= [2,2]
#OUtput: [2,3]


#Example 2:
#Input: nums1=[4,9,5]
#nums2: [9,4,9,8,4]
#Output: [4,9]


#Follow Up Questions:
#What if the given array is already srted?
#How would you optimize your algorithm?
#What if num1's size is small compard to num2's size
#? Which algorithm is better?

#WHat if elements of nums2 are sorted on disk,
#and the memory is limited such that you cannot
#load all elements into the memory
#at once?



#Ideas how to proceed:
#compare length of 2 strings
#take the longer one
#Build a frequency hash map

#Compare the elements of the second array
#with the frequency h

class Solution(object):
    def hashmap(self,num):
        #How to build a hash map??
        dict = {}
        for i in num:
            if i not in dict:
                dict[i]=1
            elif i in dict:
                dict[i]+=1
        return dict

    def intersect(self,nums1,nums2):
        common = []
        if len(nums1)>len(nums2):
            #Build hashmap
            dictionary = self.hashmap(nums1)
            for i in nums2:
                if i in dictionary and dictionary[i]!=0:
                    dictionary[i]-=1
                    if i not in common:
                        common.append(i)
        else:
            dictionary=self.hashmap(nums2)
            for i in nums1:
                if i in dictionary and dictionary[i]!=0:
                    dictionary[i]-=1
                    if i not in common:
                        common.append(i)
        return common

if __name__ == "__main__":
    nums1 = [4,9,5]
    nums2 = [9,4,9,8,4]
    output = Solution()
    print ('The intersecting elements are:',output.intersect(nums1,nums2))


    nums3 = [1,2,2,1]
    nums4 = [2,2]
    print ('The intersecting elements are:',output.intersect(nums3,nums4))

File: LeetCode/Apple/TopQuestions/Sorting_Searching/intersect.cpp

#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <set>

using namespace std;
class Solution{
public:
  vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
      if (nums1.size() > nums2.size()) {
          return intersect(nums2, nums1);
      }
      unordered_map<int, int> m;
      for (auto n : nums1)
      {
          ++m[n];
      }
      int k = 0;
      for (auto n : nums2) {
          auto it = m.find(n);
          if (it != end(m) && --it->second >= 0) {
              nums1[k++] = n;
          }
      }
      return vector(begin(nums1), begin(nums1) + k);
  }
};


class Solution_old
{
public:
  vector <int> intersect(vector<int> & nums1,vector<int>& nums2)
  {
    vector <int> common;
    map <int,int> dictionary;
    if (nums1.size()>nums2.size())
    {
      dictionary = hashmap(vector <int> & nums1);
      if ()

      dictionary[i]-=1;
    }
    else
    {
      dictionary = hashmap(vector <int> & nums2);
      dictionary[i]-=1;
    }
  return common;
  }
};


int main()
{
  vector <int> nums1 = {4,9,5};
  vector  <int> nums2 = {9,4,9,8,4};
  Solution S1;
  cout << S1.intersect(nums1,nums2) << endl;

  Solution_old S2;
  cout << "The intersection elements according to my implementation are:",S2.intersect(nums1,nums2)<<endl;
  return 0;
}
